File: 139-word-break/word-break.cpp
class Solution {
public:
    bool solve(string &s,unordered_set<string>&st,vector<int>&dp,int i){
        if(i>=s.length()) return true;

        if(dp[i]!=-1) return dp[i];
        if(st.find(s)!=st.end()) return true;
        for(int j=1;i+j<=s.length();j++){
             
             string temp=s.substr(i,j);
             if(st.find(temp)!=st.end() && solve(s,st,dp,i+j)) return dp[i]= true;
        }
        return dp[i]=false;
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<int>dp(s.length(),-1);
        unordered_set<string>st;
        for(auto i:wordDict){
            st.insert(i);
        }
        return solve(s,st,dp,0);
    }
};

File: 139-word-break/word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        HashSet<String> st = new HashSet<>(wordDict);
        int n = s.length();
        Boolean[][] memo = new Boolean[n][n + 1];
        return solve(n - 1, n, st, s, memo);
    }

    boolean solve(int i, int j, HashSet<String> st, String s, Boolean[][] memo) {
        if (i < 0) return false;
        if (i == 0) return st.contains(s.substring(i, j));

        if (memo[i][j] != null) return memo[i][j];

        if (st.contains(s.substring(i, j)) && solve(i - 1, i, st, s, memo))
            return memo[i][j] = true;

        return memo[i][j] = solve(i - 1, j, st, s, memo);
    }
}
